// engram/commons/commons.js

"use strict"




var commons = {
	cli:         { },
	contentType: { },
	random:      { },
	async:       { },
	error:       { },
	hof:         { }
}






commons.random.digits = len => {

	var out = ''

	for (var ith = 0; ith < len; ++ith) {
		out += Math.floor(Math.random( ) * 10)
	}

	return out

}

commons.contentType.isHTML = type => {
	return type === 'text/html'
}

commons.contentType.isPDF  = type => {
	return type === 'application/pdf'
}





commons.async.collectMap = (data, fn0, fn1) => {

	var recur = (aggregate, data, fn0, fn1) => {

		if (data.length === 0) {
			fn1(aggregate)
		} else {

			fn0(data.pop( ), result => {
				recur(aggregate.concat([result]), data, fn0, fn1)
			}, aggregate.length)
		}
	}

	recur([ ], data, fn0, fn1)

}





commons.cli.renameArgs = args => {

	var renamed = { }

	Object.keys(args).forEach(rawKey => {
		renamed[ rawKey.replace(/^[-][-]/g, '') ] = args[rawKey]
	})


	return renamed

}




commons.hof.objMap = (obj, fn) => {

	var out = { }

	Object.keys(obj).forEach(key => {
		out[key] = fn(key, obj[key])
	})

	return out

}




module.exports = commons
