// engram/es5/create-credentials.js
#!/usr/bin/env node


"use strict";

var is = require('is');
var pem = require('pem');
var log = require('engram/es5/logging/logger');
var constants = require('engram/es5/constants');
var fs = require('fs');

var Result = require('engram/es5/result').Result;
var Err = require('engram/es5/result').Err;

var credentialConfig = {

	days: constants.CERT_EXPIRATION,
	selfSigned: true,

	keyBitsize: constants.CERT_KEYSIZE,
	commonName: constants.CERT_COMMONNAME,
	country: constants.CERT_COUNTRY

};

var createCredentials = function createCredentials(keypath, certpath, callback) {

	log.global.info('creating https certificates.');

	createCredentials.precond(keypath, certpath, callback);

	pem.createCertificate(credentialConfig, function (err, keys) {

		if (err) {
			callback(Err(err));
		}

		var writeResult = Result.of(function () {
			fs.writeFileSync(keypath, keys.serviceKey);
			fs.writeFileSync(certpath, keys.certificate);
		});

		callback(writeResult);
	});
};

createCredentials.precond = function (keypath, certpath, callback) {

	is.always.string(keypath);
	is.always.string(certpath);
	is.always['function'](callback);
};

module.exports = createCredentials;
