// engram/es5/logger.js
#!/usr/bin/env node


"use strict";

var bunyan = require('bunyan');

var log = bunyan.createLogger({
	name: 'engram',
	streams: [{
		type: 'rotating-file',
		path: 'log/engram.log',
		period: '1d',
		count: 5
	}, {
		stream: process.stdout
	}]
});

var stats = bunyan.createLogger({
	name: 'stats',
	streams: [{
		type: 'rotating-file',
		path: 'log/stats.log',
		period: '1d',
		count: 5
	}, {
		stream: process.stdout
	}]
});

var security = bunyan.createLogger({
	name: 'security',
	streams: [{
		type: 'rotating-file',
		path: 'log/security.log',
		period: '1d',
		count: 5
	}, {
		stream: process.stdout
	}]
});

var jobResult = bunyan.createLogger({
	name: 'job-result',
	streams: [{
		type: 'rotating-file',
		path: 'log/job-result.log',
		period: '1d',
		count: 5
	}, {
		stream: process.stdout
	}]
});

var err = bunyan.createLogger({
	name: 'err',
	streams: [{
		type: 'rotating-file',
		path: 'log/err.log',
		period: '1d',
		count: 5
	}, {
		stream: process.stdout
	}]
});

module.exports = {
	global: log,
	stats: stats,
	security: security,
	jobResult: jobResult,
	err: err
};
