// engram/es5/routes-save.js
#!/usr/bin/env node


"use strict";

var log = require('engram/es5/logger');

var Result = require('engram/es5/result').Result;
var Ok = require('engram/es5/result').Ok;
var Err = require('engram/es5/result').Err;

var is = require('is');
var express = require('express');

var saveBookmark = require('engram/es5/save-bookmark');

var URL = require('url');
var validURL = require('valid-url');

var saveArchive = require('engram/es5/save-archive');
var saveSoftFailStatus = require('engram/es5/save-softfail');
var saveStatus = require('engram/es5/save-status');
var saveTitle = require('engram/es5/save-title');

var constants = require('engram/es5/constants');
var message = require('engram/es5/message');

var requestURL = require('engram/es5/request-url');

require('engram/es5/stats');
require('engram/es5/job-result');

var save = express.Router({
	strict: true,
	mergeParams: true
});

var isIgnored = function isIgnored(url) {

	return [/^http[:]\/\/localhost/g, /css[.]map$/g, /^javascript[:]/g].some(function (pattern) {
		return pattern.test(url);
	});
};

var parseURL = function parseURL(url) {

	return Result.of(function () {

		var parsed = URL.parse(url);

		if (!is.string(parsed.protocol)) {
			url = 'http://' + url;
		}

		return validURL.is_web_uri(url) ? Ok(url) : Err(url);
	});
};

save.use(function (req, res) {

	var pathResult = Ok(req).then(function (req) {
		return req.url;
	}).then(function (url) {
		return url.slice(1);
	}).then(parseURL);

	pathResult.onErr(function (path) {

		log.global.error({
			url: req.url,
			reqId: req.id,
			path: path
		}, 'failed to save url');

		res.status(404);
		res.end();
	});

	pathResult.onOk(function (url) {

		log.global.info({
			url: req.url,
			reqId: req.id
		}, 'GET /' + url);

		if (isIgnored(url)) {

			log.global.info({
				url: req.url,
				reqId: req.id
			}, 'ignoring url.');

			res.sendStatus(404);
		} else {

			var time = Math.round(new Date().getTime() / 1000);

			constants.BUS.emit(message.URL_REQUESTED, {
				url: url,
				time: time,
				res: res
			});
		}
	});
});

constants.BUS.on(message.URL_REQUESTED, function (_ref) {
	var url = _ref.url;
	var time = _ref.time;
	var res = _ref.res;

	saveBookmark(url, time, function (result) {

		result.onErr(function (err) {

			log.global.error({
				message: err.message,
				stack: err.stack
			}, 'failed to save ' + url);

			if (res) {

				res.status(500);
				res.send('whoops! ' + url + ' failed.');
				res.end();
			}
		});

		result.onOk(function () {

			if (res) {
				res.status(204);
				res.end();
			}

			requestURL(url, { encoding: null }, function (reqResult) {

				reqResult.onErr(function (_ref2) {
					var uri = _ref2.uri;
					var err = _ref2.err;

					log.global.error({ message: err.message, stack: err.stack }, 'failed to save ' + uri);
				});

				reqResult.onOk(function (_ref3) {
					var res = _ref3.res;
					var body = _ref3.body;

					saveStatus({ url: url, res: res, body: body }, function (res) {
						constants.BUS.emit(message.STATUS_FINISHED, res);
					});

					saveTitle({ url: url, res: res, body: body }, function (res) {
						constants.BUS.emit(message.TITLE_FINISHED, res);
					});

					saveSoftFailStatus({ url: url, res: res, body: body }, function (res) {
						constants.BUS.emit(message.SOFTFAIL_FINISHED, res);
					});

					saveArchive({ url: url, res: res, body: body }, function (res) {
						constants.BUS.emit(message.ARCHIVE_FINISHED, res);
					});
				});
			});
		});
	});
});

module.exports = save;
