// engram/create-archive.js
#!/usr/bin/env node
"use strict";

var is = require("is");
var log = require("engram/logger");

var Result = require("engram/result").Result;
var Ok = require("engram/result").Ok;
var Err = require("engram/result").Err;

var ctype = require("content-type");
var contentTypes = require("engram/content-types");

var requestURL = require("engram/request-url");

var createArchive = function (url, callback) {

	requestURL(url, { encoding: null }, function (resResult) {

		resResult.onErr(function (err) {
			callback(Err(err));
		});

		resResult.onOk(function (_ref) {
			var res = _ref.res;
			var body = _ref.body;

			var contentType = res.headers["content-type"] || "text/html";
			var parsed = Result.of(function () {
				return ctype.parse(contentType);
			});

			parsed.onErr(function (err) {

				log.error({
					contentType: contentType,
					message: err.message,
					stack: err.stack,
					url: url
				}, "content-type parsing failed.");

				callback(Err(err));
			});

			parsed.onOk(function (contentType) {

				contentTypes.isHTML(contentType) ? createArchive.html(url, body, contentType, callback) : createArchive.fallback(url, body, contentType, callback);
			});
		});
	}, true);
};

createArchive.html = function (url, body, contentType, callback) {

	createArchive.html.precond(url, body, contentType, callback);

	var content = "<h1>hello!</h1>";

	callback(Ok({
		content: content,
		contentType: ctype.format(contentType)
	}));
};

createArchive.html.precond = function (url, body, contentType, callback) {

	is.always.string(url);
	is.always.object(body);
	is.always.object(contentType);
};

createArchive.fallback = function (url, body, contentType, callback) {

	createArchive.fallback.precond(url, body, contentType, callback);

	callback(Ok({
		content: body,
		contentType: ctype.format(contentType)
	}));
};

createArchive.fallback.precond = function (url, body, contentType, callback) {

	is.always.string(url);
	is.always.object(body);
	is.always.object(contentType);
};

module.exports = createArchive;
