// engram/daemon-title.js
#!/usr/bin/env node
"use strict";

var log = require("engram/logger");

var Result = require("engram/result").Result;
var Ok = require("engram/result").Ok;
var Err = require("engram/result").Err;

var requestURL = require("engram/request-url");
var URL = require("url");

var database = require("engram/database");
var extractTitle = require("engram/extract-title");

var findUntitledURL = function (callback) {

	var sql = "\n\tSELECT * FROM url\n\n\tWHERE url.title is NULL\n\n\tORDER BY RANDOM()\n\tLIMIT 1;\n\t";

	database["do"](function (db, callback) {

		db.all(sql, function (err, rows) {
			err ? callback(Err(err)) : callback(Ok(rows));
		});
	}, callback);
};

var updateTitle = function (url, title, callback) {

	var sql = "UPDATE url SET title=$title WHERE url.url = $url";

	database["do"](function (db, callback) {

		db.run(sql, {
			$url: url,
			$title: title
		});
	}, callback);
};

var saveTitle = function (reporter, urls) {

	urls.length === 0 ? reporter(Ok(undefined)) : extractTitle(urls[0].url, function (titleResult) {

		titleResult.onErr(function (err) {
			reporter(Err(Err));
		});

		titleResult.onOk(function (title) {
			updateTitle(urls[0].url, title, function (updateResult) {
				reporter(updateResult);
			});
		});
	});
};

var job = function (reporter) {
	findUntitledURL(function (urlResult) {

		urlResult.onErr(function (err) {
			reporter(Err(err));
		});

		urlResult.onOk(saveTitle.bind({}, reporter));

		reporter("");
	});
};

var reporter = function (url) {
	console.log(url);
};

module.exports = {
	job: job,
	reporter: reporter,
	interval: 5 * 60 * 1000
};
