// engram/moniter-status.js
#!/usr/bin/env node
"use strict";

var log = require("engram/logger");

var is = require("is");
var https = require("https");
var express = require("express");

var Result = require("engram/result").Result;
var Ok = require("engram/result").Ok;
var Err = require("engram/result").Err;

var routes = require("engram/routes");
var database = require("engram/database");
var createTables = require("engram/create-tables");

var manager = require("engram/manager");

/*
	findUnresolvedURL(callback)

	select the URLS that don't have a status code or soft-fail status.
*/

var findUnresolvedURL = function (callback) {

	var sql = "\n\tSELECT * FROM url\n\n\tWHERE url.is_soft_failer is NULL\n\tOR url.status_code is NULL\n\n\tLIMIT 1;\n\t";

	database["do"](function (db, callback) {

		db.all(sql, function (err, rows) {
			err ? callback(Err(err)) : callback(Ok(rows));
		});
	}, callback);
};

/*
	updateStatus(url, statusCode, callback)

	set the status code for a URL.
*/

var updateStatus = function (url, statusCode, callback) {

	var sql = "UPDATE url SET status_code=$statusCode WHERE url.url = $url";

	database["do"](function (db, callback) {

		db.run(sql, {
			$url: url,
			$statusCode: statusCode
		});
	}, callback);
};

/*
	saveStatus(recall, state, url)

	request a URL, and save it's to the database.
*/

var saveStatus = function (recall, state, url) {

	url.length === 0 ? recall(undefined, state.control.delay) : requestURL(url[0].url, {}, function (err, res, body) {

		updateStatus(url, res.statusCode, function (result) {
			recall(url, state.control.delay);
		});
	});
};

/*
	save a single unresolved URL to the database before recalling.
*/

var moniter = function (recall, state) {
	findUnresolvedURL(function (urlResult) {

		urlResult.then(saveStatus.bind({}, recall, state));
	});
};

/*
	report whether a url was saved or not.
*/

var reporter = function (url, state) {

	if (url.length === 0) {
		state.stats.noURL += 1;
	} else {
		state.stats.urlFound += 1;
	}

	return state;
};

var state = {
	control: {
		delay: 1000
	},
	stats: {
		noURL: 0,
		urlFound: 0
	}
};

module.exports = {
	moniter: moniter,
	reporter: reporter,
	state: state
};
