// engram/moniter-softfail.js
#!/usr/bin/env node
"use strict";

var log = require("engram/logger");

var Result = require("engram/result").Result;
var Ok = require("engram/result").Ok;
var Err = require("engram/result").Err;

var database = require("engram/database");

/*
	findUncheckedURL(callback)

	select a URL that doesn't have a soft-fail status.
*/

var findUncheckedURL = function (callback) {

	var sql = "\n\tSELECT * FROM url\n\n\tWHERE url.is_soft_failer is NULL\n\n\tLIMIT 1;\n\t";

	database["do"](function (db, callback) {

		db.all(sql, function (err, rows) {
			err ? callback(Err(err)) : callback(Ok(rows));
		});
	}, callback);
};

/*
	updateSoftFail(url, isSoftFailer, callback)

	set the soft-fail status for a URL.
*/

var updateSoftFail = function (url, isSoftFailer, callback) {

	var sql = "UPDATE url SET is_soft_failer=$isSoftFailer WHERE url.url = $url";

	database["do"](function (db, callback) {

		db.run(sql, {
			$url: url,
			$isSoftFailer: isSoftFailer
		});
	}, callback);
};

var isErrorCode = function isErrorCode(code) {
	return [403, 404, 410].indexOf(code) !== -1 || code >= 500;
};

/*
	isSoftFailer(url, callback)

	does a URL fail with a 200-status? The algorithm used is
	described "Towards an understanding of the web decay."


	* if it succeeds, the page may not be dead.
	* if it redirects, follow ten non-looping steps before declaring dead.
	* if the page succeeds:
	*     create several fake URLs from the parent url.

*/

var isSoftFailer = function (url, callback) {};

/*
	saveSoftFailStatus(recall, state, url)


*/

var saveSoftFailStatus = function (recall, state, url) {

	url.length === 0 ? recall(undefined, state.control.delay) : isSoftFailer(function (statusResult) {
		updateSoftFail(url);
	});
};

/*
	save a single unresolved URLs softfail status to the database before recalling.
*/

var moniter = function (recall, state) {
	findUnresolvedURL(function (urlResult) {

		urlResult.then(saveStatus.bind({}, recall, state));
	});
};

var state = {
	control: {
		delay: 1000
	},
	stats: {}
};

module.exports = {
	moniter: moniter,
	reporter: reporter,
	state: state
};
