// engram/save-bookmark.js
#!/usr/bin/env node
"use strict";

var log = require("engram/logger");

try {

	var Result = require("engram/result").Result;
	var Ok = require("engram/result").Ok;
	var Err = require("engram/result").Err;

	var extractTitle = require("engram/extract-title");
	var createArchive = require("engram/create-archive");
	var log = require("engram/logger");
	var database = require("engram/database");

	var request = require("request");
} catch (err) {

	log.fatal({
		message: err.message,
		stacK: err.stack }, "failed to load dependencies.");

	process.exit();
}

var requestUrl = function (url, callback) {

	var opts = {
		url: url,
		headers: {
			"User-Agent": "Mozilla/5.0 (Windows NT 6.0) AppleWebKit/537.31 (KHTML, like Gecko) Chrome/26.0.1410.43 Safari/537.31"
		}
	};

	request(opts, function (err, res, body) {

		if (err) {

			log.error({
				url: url,
				message: err.message,
				stack: err.stack
			}, "url lookup failed.");

			callback(Err(err.message));
		} else {

			log.info({ url: url }, "url lookup succeeded.");
			callback(Ok({ res: res, body: body }));
		}
	});
};

var insertBookmark = function (url, time, title) {

	database["do"](function (db) {

		var statement = db.prepare("INSERT INTO bookmarks VALUES (NULL, ?, ?, ?);");

		statement.run(url, title, time);
		statement.finalize();
	});
};

var saveTitle = function (url, time, content) {
	extractTitle(url, content, insertBookmark.bind({}, url, time));
};

var saveArchive = function (url, content) {};

var saveBookmark = function (url, time, _ref) {
	var req = _ref.req;
	var res = _ref.res;

	requestUrl(url, function (contentResult) {

		contentResult.then(saveTitle.bind({}, url, time));
		contentResult.then(saveArchive.bind({}, url));
	});
};

module.exports = saveBookmark;
