// engram/rest/parts.js

"use strict"




var is        = require('is')
var log       = require('engram/logging/logger')
var commons   = require('engram/commons/commons')
var constants = require('engram/commons/constants')



var routeUtils = { }





var placeholders = { }





placeholders.validator = (method, route, req, res, next, ...args) => {

	log.global.info({

		router: `${method.toUpperCase( )} ${route}`,
		path:   `${req.method.toUpperCase( )} ${req.path}`

	}, 'request being handled.')

	next( )

}





placeholders.preprocessor = (req, res, next, ...args) => {

	req.engram = {
		id: commons.random.digits(constants.sizes.id)
	}

	res.engram = {
		startTime: process.hrtime( )
	}

	next( )

}





placeholders.main = (req, res, next, ...args) => {
	next( )
}





placeholders.postprocessor = (req, res, next, ...args) => {

	placeholders.postprocessor.logElapsed(res)
	next( )

}

placeholders.postprocessor.logElapsed = res => {

	var elapsed      = process.hrtime(res.engram.startTime)
	var nanoseconds  = (elapsed[0] * 1e9) + elapsed[1]
	var milliseconds = Math.round(nanoseconds / 1e6, 0)

}




placeholders.sender = (req, res, next, ...args) => {

	log.global.info({ }, 'no response sent; using stubbed sender.')
	next( )

}





var parts = { }

parts.router = ({

	mergeParams   = false,
	caseSensitive = true,
	strict        = false,
	mount

}) => {

	parts.router.precond(mergeParams, caseSensitive, strict, mount)

	return {mergeParams, caseSensitive, strict, mount}

}

parts.router.precond = (mergeParams, caseSensitive, strict, mount) => {

	is.always.boolean(mergeParams)
	is.always.boolean(caseSensitive)
	is.always.boolean(strict)
	is.always.string(mount)

}





parts.route = ({

	router   = 'app',
	method,
	protocol = 'https',
	route,
	id

}) => {

	parts.route.precond(router, method, protocol, route, id)

	return {
		router,
		method,
		protocol,
		route,
		id,

		validators:     [placeholders.validator.bind({ }, method, route)],
		preprocessors:  [placeholders.preprocessor],
		main:           [placeholders.main],
		postprocessors: [placeholders.postprocessor],
		senders:        [placeholders.sender]
	}

}

parts.route.precond = (router, method, protocol, route, id) => {

	is.always.string(router)
	is.always.string(method)
	is.always.string(protocol)
	is.always.string(id)

}





module.exports = parts
