// engram/rest/route-specs.js

"use strict"




var is        = require('is')
var parts     = require('engram/rest/parts')
var constants = require('engram/commons/constants')




var specs = {
	routers: { },
	routes:  [ ]
}





specs.routers = {
	api:  parts.router({

		mount: '/api'

	}),
	app:  parts.router({

		mount: '/'

	}),
	save: parts.router({

		mount:       '/',
		strict:      true,
		mergeParams: true

	})
}




specs.routes = [

	parts.route({

		id:     constants.routes.ids.GET_BOOKMARKS,
		method: 'get',
		route: '/bookmarks'

	}),

	parts.route({

		id:     constants.routes.ids.POST_IMPORT,
		method: 'post',
		route:  '/import'

	}),

	parts.route({

		id:     constants.routes.ids.GET_BOOKMARKS_ID,
		method: 'get',
		route: '/bookmarks/:id'

	}),

	parts.route({

		id:     constants.routes.ids.DELETE_ARCHIVE_ID,
		method: 'delete',
		route: '/archive/:id'

	}),

	parts.route({

		id:     constants.routes.ids.GET_STATIC_FILES,
		method: 'get',
		route:  '/static'

	})

]






module.exports = specs
